openmc/capi/mesh.py

```diff
@@ -0,0 +1,173 @@
+_dll.openmc_get_mesh_index.argtypes = [c_int32, POINTER(c_int32)]
+_dll.openmc_get_mesh_index.restype = c_int
+_dll.openmc_get_mesh_index.errcheck = _error_handler
+
+
+class Mesh(_FortranObjectWithID):
+    """Mesh stored internally.
+
+    This class exposes a mesh that is stored internally in the OpenMC
+    library. To obtain a view of a mesh with a given ID, use the
+    :data:`openmc.capi.meshes` mapping.
+
+    Parameters
+    ----------
+    index : int
+         Index in the `meshes` array.
+
+    Attributes
+    ----------
+    id : int
+        ID of the mesh
+
+    """
+    __instances = WeakValueDictionary()
+
+    def __new__(cls, uid=None, new=True, index=None):
+        mapping = meshes
+        if index is None:
+            if new:
+                # Determine ID to assign
+                if uid is None:
+                    uid = max(mapping, default=0) + 1
+                else:
+                    if uid in mapping:
+                        raise AllocationError('A mesh with ID={} has already '
+                                              'been allocated.'.format(uid))
+
+                index = c_int32()
+                _dll.openmc_extend_meshes(1, index, None)
+                index = index.value
+            else:
+                index = mapping[uid]._index
+
+        if index not in cls.__instances:
+            instance = super().__new__(cls)
+            instance._index = index
+            if uid is not None:
+                instance.id = uid
+            cls.__instances[index] = instance
+
+        return cls.__instances[index]
+
+    @property
+    def id(self):
+        mesh_id = c_int32()
+        _dll.openmc_mesh_get_id(self._index, mesh_id)
+        return mesh_id.value
+
+    @id.setter
+    def id(self, mesh_id):
+        _dll.openmc_mesh_set_id(self._index, mesh_id)
+
+    @property
+    def dimension(self):
+        dims = POINTER(c_int)()
+        n = c_int()
+        _dll.openmc_mesh_get_dimension(self._index, dims, n)
+        return tuple(as_array(dims, (n.value,)))
+
+    @dimension.setter
+    def dimension(self, dimension):
+        n = len(dimension)
+        dimension = (c_int*n)(*dimension)
+        _dll.openmc_mesh_set_dimension(self._index, n, dimension)
+
+    @property
+    def lower_left(self):
+        return self._get_parameters()[0]
+
+    @property
+    def upper_right(self):
+        return self._get_parameters()[1]
+
+    @property
+    def width(self):
+        return self._get_parameters()[2]
+
+    def _get_parameters(self):
+        ll = POINTER(c_double)()
+        ur = POINTER(c_double)()
+        w = POINTER(c_double)()
+        n = c_int()
+        _dll.openmc_mesh_get_params(self._index, ll, ur, w, n)
+        return (
+            as_array(ll, (n.value,)),
+            as_array(ur, (n.value,)),
+            as_array(w, (n.value,))
+        )
+
+    def set_parameters(self, lower_left=None, upper_right=None, width=None):
```

Comment:
Also wondering if it was intentional to name set_parameters without an underscore and _get_parameters with an underscore (is one supposed to be private and the other public?)

Reply:
That was intentional actually. It was easiest to require that they all be set at the same time; otherwise, the Mesh object on the Fortran side would only be partially constructed if a user only assigned the lower_left but not the others which makes error checking rather difficult. However, once the parameters are set, you can retrieve them one by one, which is why I have getters as properties (that call the hidden `_get_parameters` behind the scenes).

Reply:
I see, that makes sense